persistence/nhibernate/saga-concurrency.md

```diff
@@ -16,18 +16,29 @@ redirects:
 
 
-## Default behavior
+## Default concurrency behavior
 
-As stated in [saga concurrency](/nservicebus/sagas/concurrency.md), the saga persistence system depends on the data access providing an optimistic approach to concurrency. With NHibernate this results in appending a `WHERE` clause containing all known values of saga data fields when doing `UPDATE`s. This ensures that the saga data is still in the same state as when it was read.
+The NHibernate persister by default uses pessimistic locking in combination with optimistic concurrency control by comparing entity state. Modifications only succeed when this state was equal to the previously tracked state.
 
-This approach has a downside of poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads. These threads read the same saga state and process their messages but only one can succeed persisting the new state. Other threads processing the same message will experience concurrency violation error and need to retry. 
+Pessimistic locking is achieved by performing a **SELECT ... FOR UPDATE**, see [NHibernate Chapter 12. Transactions And Concurrency](https://nhibernate.info/doc/nhibernate-reference/transactions.html)
 
-In order to overcome the problem of frequent retries the NHibernate saga persister uses **additional** pessimistic concurrency control using an `UPDLOCK` hint. A lock is created when fetching the saga instance from the database and is held till the end of the transaction blocking other threads that try fetching that particular saga. Different saga instances can still be processed concurrently.
+Please read the guidance on [saga concurrency](/nservicebus/sagas/concurrency.md) on potential improvements.
+
+
+Creating instance race condition error:
+```
+NHibernate.Exceptions.GenericADOException: could not execute batch command.[SQL: SQL not available] ---> System.Data.SqlClient.SqlException: Violation of UNIQUE KEY constraint 'UQ__OrderSag__C3905BCE71EF212B'. Cannot insert duplicate key in object 'dbo.OrderSagaData'. The duplicate key value is (e87490ba-bb56-4693-9c0a-cf4f95736e06).
+```
+
+Updating instances race condition error:
+```
+NHibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [OrderSagaData#8378bd96-8143-48b2-ae3e-aad100a37cb9] ---> NHibernate.StaleStateException: Unexpected row count: 0; expected: 1
+```
 
 
 ## Explicit version
 
-The `RowVersion` attribute can be used to explicitly denote a property that should be used for optimistic concurrency control
+The `RowVersion` attribute can be used to explicitly denote a property that should be used for optimistic concurrency control.
```

Comment:
```suggestion
The `RowVersion` attribute can be used to explicitly denote a property that should be used for optimistic concurrency control. An update will then compare against this single property instead of comparing it against the previous state of all properties which results in a more efficient comparison.
```